### PR TITLE
[CALCITE-4512] Optimize when GROUP BY or HAVING alias should be substituted with select item

### PR DESCRIPTION
**Pull request by hannerwang**
for GROUP BY alias substituting:
1. if the GROUP BY expression is an identifier.
2. CUBE, ROLLUP, GROUPING SETS, ROW top level argument behave in the same way with GROUP BY expression.

for HAVING alias substituting:
1. any except AGGREGATE or its arguments .

### CODE REVIEW

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 1]
```diff
@@ -86,16 +86,19 @@ LogicalAggregate(group=[{0}])
...
+    LogicalJoin(condition=[true], joinType=[inner])
+      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+      LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
 ]]>
         </Resource>
         <Resource name="sql">
-            <![CDATA[select empno, ename as deptno from emp group by empno, deptno]]>
+            <![CDATA[select empno, case when emp.deptno is null then dept.deptno else emp.deptno
+end as deptno from emp, dept group by empno, deptno]]>
         </Resource>
```
[the remark this answers is not in the document]
Comment (hannerwang) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:102]: thanks for reviewing, the strategy that alias will be substituted has changed, so the case should be changed also.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 2]
```diff
@@ -6669,9 +6729,95 @@ private SqlNode nthSelectItem(int ordinal, final SqlParserPos pos) {
...
+        } else {
+          expandingTopNodes.put(operand, Boolean.TRUE);
+        }
+      }
+    }
+
+    /*
+     * check if top level node should be expanded as alias.
+     */
+    void checkExpandedAsAlias(SqlNode sqlNode) {
```
[the remark this answers is not in the document]
Comment (hannerwang) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:6759]: do you mean the 'check' in comment?

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 3]
```diff
@@ -371,7 +372,8 @@ public final Sql sql(String sql) {
   }
 
   @Test void testAliasInHaving() {
-    sql("select count(empno) as e from emp having e > 1")
+    sql("select ename, count(empno) as e, count(deptno) as d from emp"
+        + " group by ename having e > 1 and count(deptno) > 2")
```
[the remark this answers is not in the document]
Comment (hannerwang) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:376]: yes, that conforms SOLID,  but the commit is fixing the existing feature, so I choose to change the obsolete tests.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:6748]: Yes~~

[the remark this answers is not in the document]
Comment (hannerwang) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:6748]: thanks, I have fixed it.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 4]
```diff
@@ -142,14 +145,15 @@ LogicalAggregate(group=[{}], EXPR$0=[COUNT()])
...
-LogicalFilter(condition=[>($0, 1)])
-  LogicalAggregate(group=[{}], E=[COUNT()])
-    LogicalProject(EMPNO=[$0])
-      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+LogicalProject(ENAME=[$0], E=[$1], D=[$1])
+  LogicalFilter(condition=[>($1, 2)])
+    LogicalAggregate(group=[{0}], E=[COUNT()])
+      LogicalProject(ENAME=[$1], EMPNO=[$0], DEPTNO=[$7])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
 ]]>
```
Comment (chunweilei) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:153]: The plan seems not right. There are two aggregates in sql, but only one in plan. Do I miss something?

Comment (hannerwang) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:153]: it's the result of reducing COUNT aggregation whose args type has NOT NULL constraint to be count(*). In this case, EMPNO and DEPTNO are integer and not null, so count(empno) and count(deptno) can be reduced as just one count(*), furthermore the count(*) > 1 and count(*) > 2 can be reduced to count(*) > 2.

**Comment by Aaaaaaron** [comment]
The commit message not in the proper way.

**Pull request closed without merging**